File: src/main/java/org/hisp/dhis/expression/UnaryOperator.java
package org.hisp.dhis.expression;

import java.util.List;

import static java.lang.Math.negateExact;

public enum UnaryOperator implements Typed{

    PLUS("+", ValueType.NUMBER),
    MINUS("-", ValueType.NUMBER),
    NOT("!", ValueType.BOOLEAN),
    DISTINCT("distinct", ValueType.SAME);

    private final String symbol;
    private final ValueType returnType;

    UnaryOperator(String symbol, ValueType returnType) {
        this.symbol = symbol;
        this.returnType = returnType;
    }

    public String getSymbol() {
        return symbol;
    }

    @Override
    public ValueType getValueType() {
        return returnType;
    }

    /**
     * Avoid defensive copy when finding operator by symbol
     */
    private static final List<UnaryOperator> VALUES = List.of(values());

    static UnaryOperator fromSymbol(String symbol) {
        return VALUES.stream().filter(op -> op.symbol.equals(symbol)).findFirst().orElseThrow();
    }

    public static Number negate(Number value) {
        if (value == null)
        {
            return null;
        }
        return value instanceof Integer ? negateExact(value.intValue()) : -value.doubleValue();
    }
}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/ast/UnaryOperator.kt
package org.hisp.dhis.lib.expression.ast

import org.hisp.dhis.lib.expression.spi.ValueType

enum class UnaryOperator(val symbol: String, private val returnType: ValueType) : Typed {

    PLUS("+", ValueType.NUMBER),
    MINUS("-", ValueType.NUMBER),
    NOT("!", ValueType.BOOLEAN),
    DISTINCT("distinct", ValueType.SAME);

    override fun getValueType(): ValueType {
        return returnType
    }

    companion object {
        /**
         * Avoid defensive copy when finding operator by symbol
         */
        private val VALUES = listOf(*entries.toTypedArray())

        fun fromSymbol(symbol: String): UnaryOperator {
            return if ("not" == symbol) NOT else VALUES.stream().filter { op: UnaryOperator -> op.symbol == symbol }
                .findFirst().orElseThrow()
        }

        fun negate(value: Number?): Number? {
            if (value == null) {
                return null
            }
            return if (value is Int) Math.negateExact(value.toInt()) else -value.toDouble()
        }
    }
}
